src/index/compat.rs

```diff
@@ -0,0 +1,165 @@
+use crate::gucs::planning::ENABLE_PGVECTOR_COMPATIBILITY;
+use libc::c_void;
+use pgrx::pg_sys::pfree;
+use pgrx::pg_sys::AsPgCStr;
+use std::collections::HashMap;
+use std::ffi::CStr;
+
+unsafe fn swap_destroy<T>(target: &mut *mut T, value: *mut T) {
+    let ptr = *target;
+    *target = value;
+    if !ptr.is_null() {
+        pfree(ptr.cast());
+    }
+}
+
+pub unsafe fn pgvector_stmt_rewrite(pstmt: *mut pgrx::pg_sys::PlannedStmt) {
+    let enabled = ENABLE_PGVECTOR_COMPATIBILITY.get();
+    if !enabled {
+        return;
+    }
+    let utility_statement = (*pstmt).utilityStmt;
+    let is_index = pgrx::is_a(utility_statement, pgrx::pg_sys::NodeTag::T_IndexStmt);
+    if is_index {
+        let istmt: *mut pgrx::pg_sys::IndexStmt = utility_statement.cast();
+        if istmt.is_null() {
+            return;
+        }
+        let method = CStr::from_ptr((*istmt).accessMethod).to_str();
+        if method == Ok("hnsw") || method == Ok("ivfflat") {
+            rewrite_type_options(istmt, method.unwrap());
+            rewrite_opclass(istmt);
+            swap_destroy(&mut (*istmt).accessMethod, "vectors".as_pg_cstr());
+        }
+    }
+}
+
+unsafe fn rewrite_type_options(istmt: *mut pgrx::pg_sys::IndexStmt, method: &str) {
+    let original = vec_from_list((*istmt).options);
+    let opts = options_from_vec(original);
+    match method {
+        "hnsw" => {
+            let m = opts
+                .get("m")
+                .unwrap_or(&String::from("16"))
+                .parse::<u32>()
+                .unwrap();
+            let ef_construction = opts
+                .get("ef_construction")
+                .unwrap_or(&String::from("64"))
+                .parse::<usize>()
+                .unwrap();
+            let arg = pgrx::pg_sys::makeString(
+                format!(
+                    "[indexing.hnsw]\nm = {}\nef_construction = {}",
+                    m, ef_construction
+                )
+                .as_pg_cstr(),
+            );
+            let elem = pgrx::pg_sys::makeDefElem("options".as_pg_cstr(), arg as _, -1);
+            swap_destroy(&mut (*istmt).options, list_from_vec(vec![elem]));
+        }
+        "ivfflat" => {
+            let nlist = opts
+                .get("list")
+                .unwrap_or(&String::from("100"))
+                .parse::<u32>()
+                .unwrap();
+            let arg =
+                pgrx::pg_sys::makeString(format!("[indexing.ivf]\nnlist = {}", nlist).as_pg_cstr());
+            let elem = pgrx::pg_sys::makeDefElem("options".as_pg_cstr(), arg as _, -1);
+            swap_destroy(&mut (*istmt).options, list_from_vec(vec![elem]));
+        }
+        _ => {}
+    }
+}
+
+unsafe fn rewrite_opclass(istmt: *mut pgrx::pg_sys::IndexStmt) {
+    let elems = vec_from_list::<pgrx::pg_sys::IndexElem>((*istmt).indexParams);
+    if elems.is_empty() {
+        return;
+    }
+    for e in elems {
+        let opclass_name = vec_from_list::<c_void>((*e).opclass)
```

Comment:
The length can be $0$ if a user create a operator class and set it default so that they could do not have to write operator class in `CREATE INDEX`.

Reply:
I would skip it to not disturb other index opclass.